File: 7_kyu/Sum_of_a_sequence.go
package kata


// Calculates the sum of sequence
//    start, start+step, start+2*step, ..., x.
// with x <= end
func SequenceSum(start, end, step int) int {
  // Use Gauss sum formula
  n := (step+end-start)/step
  if n <= 0 {
    // Case start > end
    return 0
  }
  return start * n + step*(((n-1)*n) / 2)
}

File: 7_kyu/Sum_of_a_sequence.c
/**
 *  Calculates the sum of a sequence of integers.
 *  start: first number
 *  end: all numbers in the sequence must be less than or equal to end.
 *  step: step (difference) between the numbers in the sequence.
 *  return: sum or 0 on invalid arguments.
 */
unsigned sequence_sum(unsigned start, unsigned end, unsigned step)
{
    // error exit on invalid arguments
    if (end < start || step <= 0) return 0;
    // use long for intermediate calculations to avoid overflow
    // first number in the sequence
    const unsigned long first = start;
    // count of numbers in the sequence
    const unsigned long n = (end-start)/step + 1;
    // last number in sequence less than or equal to end
    const unsigned long last = first + (n-1)*step;
    // extended Gauss sum formula
    return ((first+last)*n)/2;
}
